### PR TITLE
Feature/marshal json refactor

### LINKED ISSUE #801: MarshalJSON functions must use defined struct instead of interface{}
## Rationale

Today as you can see at https://github.com/hermeznetwork/hermez-node/blob/51d4dd5390df5f99eb746e315c751194fd700d37/db/historydb/views.go#L64 the Marshal function create a generic struct in runtime. This does not allow third-party software, such as a SDK or any other software, to reuse hermez-node code and assure the struct it is being used is the same used on hermez-node and keep the compatibility with confidence. 

Also, is not idiomatic to use interface{} in Go. Always is preferable to define a struct and in this case seems to do not have a reason to not use a defined struct.

This change must be applied in all MarshallJSON in historydb package

## Implementation

Create a struct for the data format is going to be returned by API instead of interface{}

### PR DESCRIPTION
**Pull request by mfcastellani**
Closes #801

## Rationale

Today as you can see at https://github.com/hermeznetwork/hermez-node/blob/51d4dd5390df5f99eb746e315c751194fd700d37/db/historydb/views.go#L64 the Marshal function create a generic struct in runtime. This does not allow third-party software, such as a SDK or any other software, to reuse hermez-node code and assure the struct it is being used is the same used on hermez-node and keep the compatibility with confidence. 

Also, is not idiomatic to use interface{} in Go. Always is preferable to define a struct and in this case seems to do not have a reason to not use a defined struct.

This change create structs and remove the interface use in the historydb package. 

### How to test?

`make test`

## Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Include tests - since PR is just a refactor tests are already available
- [X] Respect code style and lint
- [X] Update swagger file (if needed)
- [X] Update documentation (*.md) (if needed)

### CODE REVIEW

File: db/historydb/views.go [hunk 1]
```diff
@@ -12,6 +12,80 @@ import (
...
+	ToForgeL1TxsNum          *int64              `json:"toForgeL1TransactionsNum"`
+	UserOrigin               *bool               `json:"userOrigin"`
+	DepositAmount            *apitypes.BigIntStr `json:"depositAmount"`
+	AmountSuccess            bool                `json:"amountSuccess"`
+	DepositAmountSuccess     bool                `json:"depositAmountSuccess"`
+	HistoricDepositAmountUSD *float64            `json:"historicDepositAmountUSD"`
+	EthereumBlockNum         int64               `json:"ethereumBlockNum"`
+}
+
+type l2infoJSON struct {
```
Comment (jeffprestes) [inline db/historydb/views.go:83]: @mfcastellani , could you make these new structs public? This way we can also reuse them in SDK or external libraries.

**Review by jeffprestes — CHANGES_REQUESTED** [review]
As asked in the issue, could you please make these new structs public? All other changes looks good to me.

**Review by arnaubennassar — DISMISSED** [review]
I've been working in a similar refactor recently, and I wanted to share a few reflections and ideas on this topic (just to be clear, I'm not asking for adding this on this PR):

- We should add Unmarshalers and make structs public, to make it easy for other projects to integrate
- Once we've public structs that are able to marshal/unmarshal in the api format, they shouldn't use the `apitypes` types. Example: `EthAddr          apitypes.HezEthAddr` should be `EthAddr          ethCommon.Address`. Reasoning: until now we only cared to fulfill the API format, as it was the main duty, but this is very bad for reusing code on the client side, people don't want tor receive a weird string  but an actual Ethereum address (in this example, but idx, bjj, .... also aplies). To do so I suggest hiding this intermediary structs inside the marhsal/unmarshal functions [as I did here](https://github.com/hermeznetwork/hermez-node/pull/841/files#diff-6bb7ed027df678ce6822860d0344e836821a8c7e6ab979291dad26de7e48051aR444)
- We should try to use the `common` package as much as possible to avoid having similar structs that represent the same but require extra parsing and similar boring code. Although this is impossible in many cases due to the difference in content from the API objects and the common structs.
- I'd like to reconsider the overall architecture: to me it's a bit ugly to have things that are very specific to the api package in the db packages

**Comment by tclemos** [comment]
> * I'd like to reconsider the overall architecture: to me it's a bit ugly to have things that are very specific to the api package in the db packages

I've seen problems like this before and having a project architecture well defined can help. 

The Hexagonal Architecture is an Architecture that aims to have the core domain isolated from the external dependencies like http, db, infrastructure in general, etc.

Please, take a look in this article and see if it helps: https://medium.com/@matiasvarela/hexagonal-architecture-in-go-cfd4e436faa3

**Review by jeffprestes — APPROVED** [review]

**Pull request merged**